### ISSUE TITLE
Performance of EntityFramework IQueryable

### ISSUE THREAD

**Issue by miika1006** [NonFactors/MVC5.Grid.Web#6]
I use EntityFramework to query database. Database table has about 135000 rows, I return the results to the Grid as IQueryable<MyObject> which is not enumerated to list, but is a Linq to sql clause.

Then when I create the grid in ASP.NET MVC View and make it Pageable so that I wont have to show all the 135000 rows. It takes from 30 seconds to 1 minute for the Grid to render 10 rows of the results.

In _Grid.cshtml view, when the code runs
@if(Model.Rows.Any())  <-- this takes a long time.

I suspect it must enumerate or get all the 135000 rows from database before the paging happens?

I tried to look at the source code but I couldn't figure out how or where it gets the 10 rows I need in my view and not all the rows from database, and then filter out the paging rows.

My Grid is defined in view like this:
```C#
@model IQueryable<PrintLog>

@(Html
    .Grid(Model)
    .Build(columns =>
    {
        columns.Add(model => model.TimeStamp).Titled(Resources.TIMESTAMP);
})
.Empty(Resources.NO_ROWS)
    .Pageable(pager =>
    {
        pager.PagesToDisplay = 6;
        pager.RowsPerPage = 10;
    })
.Named("PrintLogGrid")
```

**Comment by Muchiachio** [comment]
I would recommend logging all EntityFramework queries to see why it happens, maybe `COUNT` query takes a long time for some reason, or query gets enumerated somewhere.

**Comment by miika1006** [comment]
I got the sql query that the EntityFramework generates, and used it manually in SQL Management studio, the sql alone is not that slow.

In the source code it uses IEnumerable parameter, I'm providin IQueryable to it.. is it possible that if in the source code it add some linq conditions (like .Where() or .Count()) to the IEnumerable parameter, the EntityFramework queries all the data from database and then adds the conditions to it.

Could it work if it was IQueryable all the way, untill the data is finally requested in the end using .ToList() for example

**Comment by Muchiachio** [comment]
It does work as `IQueryable` all the way, by checking EntityFramework query log you can find which query takes all the time and run it against SQL Management studio.

**Comment by miika1006** [comment]
I did that, and the sql query I got, was not that slow when ran it in SQL Management studio. So I thought that it must be Entity Framework or the Grid. I tried again this morning with a plain EntityFramework query of the table, without any conditions and it worked great.

I found out that the rendering was slow, because I had a **Group by** clause in my Linq. So EntityFramework was the reason it was so slow. I'll think of another way to show my data than group by and it will be ok.

This issue is not solved,
Thank you for your help and fast responses.
